src/Cli/Microsoft.DotNet.Cli.Utils/TypoCorrection.cs

```diff
@@ -7,40 +7,101 @@
         /// <returns>The enumerator to tokens similar to <paramref name="currentToken"/>.</returns>
         public static IEnumerable<string> GetSimilarTokens(IEnumerable<string> possibleTokens, string currentToken, int maxLevenshteinDistance = 3)
         {
-            int? bestDistance = null;
-            return possibleTokens
-                .Select(possibleMatch => (possibleMatch, distance: GetDistance(currentToken, possibleMatch)))
-                .Where(tuple => tuple.distance <= maxLevenshteinDistance)
-                .OrderBy(tuple => tuple.distance)
-                .ThenByDescending(tuple => GetStartsWithDistance(currentToken, tuple.possibleMatch))
-                .TakeWhile(tuple =>
+            var minCurrentTokenLength = 3;
+            var maxNumberOfSuggestions = 10;
+
+            var numberOfSuggestions = 0;
+            var currentTokenLength = currentToken.Length;
+            var possibleSuggestions = possibleTokens.Select((string possibleMatch) => new Suggestion(possibleMatch, possibleMatch.Length));
+
+            var matchByStartsWith = possibleSuggestions
+               .Where(s => s.PossibleMatch.StartsWith(currentToken))
+               .OrderBy(s => s.Distance)
+               .ToList();
+
+            numberOfSuggestions += matchByStartsWith.Count;
+            if (numberOfSuggestions >= maxNumberOfSuggestions)
+            {
+                return matchByStartsWith
+                    .Take(maxNumberOfSuggestions)
+                    .Select(s => s.PossibleMatch);
+            }
+
+            var matchByContains = new List<Suggestion>();
+            if (currentToken.Length >= minCurrentTokenLength)
+            {
+                matchByContains = possibleSuggestions
+                    .Where(s => !matchByStartsWith.Contains(s)
+                        && s.PossibleMatch.Contains(currentToken)
+                        && s.Distance - currentTokenLength <= maxLevenshteinDistance)
+                    .OrderBy(s => s.Distance)
+                    .ToList();
+
+                numberOfSuggestions += matchByContains.Count;
+                if (numberOfSuggestions >= maxNumberOfSuggestions)
                 {
-                    (string _, int distance) = tuple;
-                    bestDistance ??= distance;
-                    return distance == bestDistance;
-                })
-                .Select(tuple => tuple.possibleMatch);
+                    return matchByStartsWith
+                        .Concat(matchByContains)
+                        .Take(maxNumberOfSuggestions)
+                        .Select(s => s.PossibleMatch);
+                }
+            }
+
+            var matchByLevenshteinDistance = possibleSuggestions
+                .Select(s => new Suggestion(s.PossibleMatch, GetDistance(s.PossibleMatch, currentToken)))
+                .Where(s => !(matchByStartsWith.Contains(s) || matchByContains.Contains(s)) && s.Distance <= maxLevenshteinDistance)
```

Comment:
The first part of the condition `!(matchByStartsWith.Contains(s) || matchByContains.Contains(s))` should be moved before the `Select` to save us from calculating distance for strings that will not be used.